src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/DataSetCandidateService.cs

```diff
@@ -44,19 +44,27 @@ internal class DataSetCandidateService(
         Guid releaseVersionId,
         CancellationToken cancellationToken)
     {
-        return await contentDbContext.ReleaseFiles
+        var releaseFilesByFileId = await contentDbContext.ReleaseFiles
             .AsNoTracking()
             .Where(rf => rf.ReleaseVersionId == releaseVersionId)
             .Where(rf => rf.File.Type == FileType.Data)
             .Where(rf => rf.File.PublicApiDataSetId == null)
             .Where(rf => rf.File.ReplacedById == null)
             .Where(rf => rf.File.ReplacingId == null)
+            .ToDictionaryAsync(rf => rf.FileId, cancellationToken: cancellationToken);
+
+        return (await contentDbContext.DataImports
+            .Where(di => di.Status == DataImportStatus.COMPLETE)
+            .Where(di => releaseFilesByFileId.Keys.Contains(di.FileId))
+            .Select(di => di.FileId)
+            .ToListAsync(cancellationToken: cancellationToken))
+            .Select(fileId => releaseFilesByFileId[fileId])
             .Select(rf => new DataSetCandidateViewModel
             {
                 ReleaseFileId = rf.Id,
                 Title = rf.Name!,
             })
             .OrderBy(rf => rf.Title)
-            .ToListAsync(cancellationToken: cancellationToken);
+            .ToList();
```

Comment:
I initially found this a bit confusing to read because my first thought was 'why is the source of the candidates result originating from `DataImports`?'.

I understand it now but just wondering if you think it would be more readable if the file id's with data imports list was assigned a variable, something like this?

```csharp
var fileIdsWithCompletedImports = await contentDbContext.DataImports
            .Where(di => di.Status == DataImportStatus.COMPLETE)
            .Where(di => releaseFilesByFileId.Keys.Contains(di.FileId))
            .Select(di => di.FileId)
            .ToListAsync(cancellationToken: cancellationToken);

return fileIdsWithCompletedImports 
            .Select(fileId => releaseFilesByFileId[fileId])
            .Select(rf => new DataSetCandidateViewModel
            {
                ReleaseFileId = rf.Id,
                Title = rf.Name!,
            })
            .OrderBy(rf => rf.Title)
            .ToList();
```

Reply:
Fair point, but also I realised that there's a more efficient way to do this using a `Join` rather than two queries. The code now looks like this:

```cs
return await contentDbContext.ReleaseFiles
    .Where(rf => rf.ReleaseVersionId == releaseVersionId)
    .Where(rf => rf.File.Type == FileType.Data)
    .Where(rf => rf.File.PublicApiDataSetId == null)
    .Where(rf => rf.File.ReplacedById == null)
    .Where(rf => rf.File.ReplacingId == null)
    .Join(
        contentDbContext.DataImports,
        rf => rf.FileId,
        di => di.FileId,
        (rf, di) => new { ReleaseFile = rf, DataImport = di }
    )
    .Where(tuple => tuple.DataImport.Status == DataImportStatus.COMPLETE)
    .Select(tuple => tuple.ReleaseFile)
    .Select(rf => new DataSetCandidateViewModel
    {
        ReleaseFileId = rf.Id,
        Title = rf.Name!,
    })
    .OrderBy(rf => rf.Title)
    .ToListAsync(cancellationToken: cancellationToken);
```

How's this look to you?

Reply:
Looks great to me, and I saw you added `AsNoTracking()` too 🙂 .